velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#define XXH_INLINE_ALL
+
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/external/xxhash.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t hashInteger(const T& value) {
+  int32_t distance = 31;
+  int64_t num = value * PRIME64_2;
+  int64_t unsignedShift = ((uint64_t)num) >> -distance;
+  int64_t rotateLeft = (num << distance) | unsignedShift;
+  return rotateLeft * PRIME64_1;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t
+hashInteger(const DecodedVector& vector, vector_size_t row) {
+  return hashInteger(vector.valueAt<T>(row));
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+hashStringView(const DecodedVector& vector, vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
+  return hash;
+}
+} // namespace
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BIGINT>(vector_size_t row) {
+  return hashInteger<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::INTEGER>(vector_size_t row) {
+  return hashInteger<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::SMALLINT>(vector_size_t row) {
+  return hashInteger<int16_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TINYINT>(vector_size_t row) {
+  return hashInteger<int8_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  return hashInteger(vector_->valueAt<int64_t>(row));
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return hashInteger(vector_->valueAt<Date>(row).days());
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARCHAR>(vector_size_t row) {
+  return hashStringView(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARBINARY>(vector_size_t row) {
+  return hashStringView(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::REAL>(vector_size_t row) {
+  return hashInteger(vector_->valueAt<int32_t>(row));
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return hashInteger<int64_t>(*vector_.get(), row);
+}
```

Comment:
Timestamp is a struct with 2 fields. It doesn't seem valid to cast it to int64_t. I wonder if the right logic is to call Timestamp::toMillis().

```
  int64_t seconds_;
  uint64_t nanos_;
```